crates/sui-core/src/gateway_state.rs

```diff
@@ -1391,6 +1403,32 @@ where
         Ok(data)
     }
 
+    async fn pay(
+        &self,
+        signer: SuiAddress,
+        input_coins: Vec<ObjectID>,
+        recipients: Vec<SuiAddress>,
+        amounts: Vec<u64>,
+        gas: Option<ObjectID>,
+        gas_budget: u64,
+    ) -> Result<TransactionData, anyhow::Error> {
+        let used_coins = BTreeSet::from_iter(input_coins.iter().cloned());
+        let gas = self
+            .choose_gas_for_address(signer, gas_budget, gas, used_coins)
+            .await?;
+        let handles: Vec<_> = input_coins
+            .iter()
+            .map(|id| self.get_object_ref(id))
+            .collect();
+        let coins = join_all(handles)
+            .await
+            .into_iter()
+            .map(|c| c.unwrap())
+            .collect();
+        let data = TransactionData::new_pay(signer, coins, recipients, amounts, gas, gas_budget);
```

Comment:
What ensures that `gas` is not part of `coins`?

Reply:
if gas object is not specified, `choose_gas_for_address` will avoid using coin in `coins` as the gas object, but you are right if user explicitly pass a coin both to `coins` and as `gas`, it can be problematic, adding a check now. For this case, presumably we should redirect it to `paySui`